ccp/src/hashrate/hashrate_collector.rs

```diff
@@ -0,0 +1,240 @@
+/*
+ * Copyright 2024 Fluence Labs Limited
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+use serde::Deserialize;
+use serde::Serialize;
+use std::collections::HashMap;
+use std::time::Duration;
+use tokio::time::Instant;
+
+use ccp_shared::types::EpochParameters;
+use ccp_shared::types::LogicalCoreId;
+
+use super::hashrate_record::HashrateCUEntryVariant;
+use super::hashrate_record::HashrateCURecord;
+
+#[derive(Clone, Debug)]
+pub(crate) struct HashrateCollector {
+    status: CollectorStatus,
+    entries: HashMap<LogicalCoreId, CUHashrateRaw>,
+}
+
+#[derive(Copy, Clone, Debug, PartialEq, Eq)]
+pub(crate) enum CollectorStatus {
+    Started {
+        started_time: Instant,
+        epoch: EpochParameters,
+    },
+    JustCreated,
+}
+
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub(crate) struct CUHashrateRaw {
+    cache_creation: ParameterStatus<Duration>,
+    dataset_initialization: Vec<Duration>,
+    pow_duration: ParameterStatus<Duration>,
+    hashes_checked: u64,
+    proofs_found: u64,
+}
+
+#[derive(Copy, Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
+pub(crate) enum ParameterStatus<T> {
+    Measured(T),
+    NotMeasured,
+}
```

Comment:
```suggestion
#[derive(Copy, Clone, Debug, PartialEq, Eq, Serialize, Deserialize, Default)]
pub(crate) enum ParameterStatus<T> {
    Measured(T),
    #[default]
    NotMeasured,
}
```

You will be able to derive Debug on `CUHashrateRaw` with this too.